### PR TITLE
Only change documents in solution

### PR DESCRIPTION
**Pull request by Cosifne**
So according to the dump:
```
Microsoft_CodeAnalysis_Workspaces_ni!Microsoft.CodeAnalysis.Shared.Extensions.ISolutionExtensions.GetRequiredDocument(Microsoft.CodeAnalysis.Solution, Microsoft.CodeAnalysis.DocumentId)+0x2ab3ada
Microsoft_CodeAnalysis_Features_ni!Microsoft.CodeAnalysis.ChangeNamespace.AbstractChangeNamespaceService`3+<>c__DisplayClass26_0[[System.__Canon, mscorlib],[System.__Canon, mscorlib],[System.__Canon, mscorlib]].<ChangeNamespaceInSingleDocumentAsync>b__4(System.Linq.IGrouping`2<Microsoft.CodeAnalysis.DocumentId,LocationForAffectedSymbol<System.__Canon,System.__Canon,System.__Canon>>)+0x59
```
It indicates in some cases, [AbstractChangeNamespaceService.cs](https://github.com/dotnet/roslyn/compare/release/dev17.8...Cosifne:dev/shech/FilterDocumentsInSolution?expand=1#diff-1258b1f4414057819e82f3f7fec71f26796931c7d429b45c30a98ed85ae76e8f), line 478
`solutionWithChangedNamespace.GetRequiredDocument(refInOneDocument.Key)` would throw

However, we don't know the repro case : ( 
So there is no test could cover it

### CODE REVIEW

File: src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractChangeNamespaceService.cs [hunk 1]
```diff
@@ -470,7 +470,7 @@ private static SyntaxNode CreateImport(SyntaxGenerator syntaxGenerator, string n
                 .ConfigureAwait(false);
             var solutionWithChangedNamespace = documentWithNewNamespace.Project.Solution;
 
-            var refLocationGroups = refLocationsInOtherDocuments.GroupBy(loc => loc.Document.Id);
+            var refLocationGroups = refLocationsInOtherDocuments.Where(loc => solutionWithChangedNamespace.ContainsDocument(loc.Document.Id)).GroupBy(loc => loc.Document.Id);
```
Comment (genlu) [inline src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractChangeNamespaceService.cs:473]: So either FindReferences returns a reference in a Document that's not a part of the solution (seems implausible, but would be bad if true), or somehow `FixDeclarationDocumentAsync` caused the referencing document to be removed (a quick glance suggests this is not the case).

Could we add some NFW to collect more information?

Comment (Cosifne) [reply src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractChangeNamespaceService.cs]: Added

**Comment by Cosifne** [comment]
Tag @genlu  and @jasonmalinowski  to have a second review

**Review by genlu — APPROVED** [review]

File: src/Features/Core/Portable/CodeRefactorings/SyncNamespace/SyncNamespaceDocumentsNotInSolutionException.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    internal class SyncNamespaceDocumentsNotInSolutionException(ImmutableArray<DocumentId> documentIds) : Exception
+    {
+        private readonly ImmutableArray<DocumentId> _documentIds = documentIds;
+
+        public override string ToString()
+        {
+            using var _ = PooledStringBuilder.GetInstance(out var builder);
+            foreach (var documentId in _documentIds)
+            {
+                builder.AppendLine($"{documentId.GetDebuggerDisplay()}, IsSourcedGeneratedDocument: {documentId.IsSourceGenerated}");
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/CodeRefactorings/SyncNamespace/SyncNamespaceDocumentsNotInSolutionException.cs:20]:
```suggestion
                builder.AppendLine($"{documentId.GetDebuggerDisplay()}, IsSourceGeneratedDocument: {documentId.IsSourceGenerated}");
```

**Review by jasonmalinowski — APPROVED** [review]

**Pull request merged**